### DISCUSSION TITLE
Switch to anyio (at least internally)

### DISCUSSION THREAD

**Discussion by frederikaalund** [empicano/aiomqtt#44] [Ideas]
Back in October 2020, I discovered [anyio](https://github.com/agronholm/anyio). It's basically a wrapper around several event loops: asyncio, trio, curio. This way, our library could cover more ground *and* benefit from [structured concurrency](https://en.wikipedia.org/wiki/Structured_concurrency).

Fill disclosure: I'm a proponent of structured concurrency. This is something that trio implements really well. In an ideal world, we would have trio in the Python standard library instead of asyncio. Unfortunately, this is not how history wanted it (trio came too late to the party). Today, asyncio is the de facto event loop implementation in Python.

Around mid-2020, I tried to take some of the ideas from structured concurrency and implement them on top of asyncio. I wanted the best of both worlds: Structured concurrency on top of the de facto event loop implementation.
Turns out I wasn't the only one with this idea: anyio beat me too it. anyio takes the ideas of structured concurrency and makes it available on top of asyncio (and other event loops).

I was really excited to find anyio and I quickly tried to port asyncio-mqtt to anyio. Unfortunately, I rant into some [show stoppers in the anyio API](https://github.com/agronholm/anyio/issues/167) at the time.

Recently (like 3 days ago) anyio [fixed those show stoppers](https://github.com/agronholm/anyio/issues/167#issuecomment-784559598). The fix is still only available on master but it should land in a release soonish.

Sooo, what do you (the maintainers/community) think? Is structured concurrency a fad or the next big thing? Do we want to add anyio as a dependency (along with its dependencies)? Should we instead keep asyncio-mqtt lean and stick to the single-dependency (paho-mqtt)?

Let me know. :)

**Comment by flyte** [comment]
Typical! I just feel like I'm getting comfortable with asyncio and another paradigm shows up :grin: 

Without delving deeper into the details, it's hard for me to have a valid opinion, but from a surface level it sounds like a good idea.

On a personal level, I'm interested in learning how to make async easier and less error-prone, as well as catching up with the state of the art. I'm wary of using asyncio-mqtt as a whipping-boy to learn it with, but at the same time, there's no better way to learn something than by actually using it.

**Reply by flyte** [reply]
Also, what would this mean for the public API of the client? Would it require users to change how they use it, or would the change be invisible, and only internal as you allude to in the discussion topic?

**Reply by frederikaalund** [reply]
I can definitely recommend to read up a bit on structured concurrency. I think we'll see that term used a lot more in the future. :))

> Would it require users to change how they use it, or would the change be invisible, and only internal as you allude to in the discussion topic?

Good question. :+1: I *think* that we can pull this of without breaking the API. I have a work-in-progress anyio-based port lying around somewhere. I basically took asyncio-mqtt and changed the internal event loop logic. I never finished it though, due to the show stoppers mentioned in the original post.

In any case, it *is* a major internal rework and there *will* be issues due to this. If we go in this direction, we'll just have to fix those as we go along.

I'm mostly concerned about bringing anyio in as a dependency (along with its set of dependencies). What will people think of that? Is asyncio-mqtt no longer "lean" now that it has multiple dependencies?

Also, suddenly the name "asyncio-mqtt" doesn't really apply anymore. The code would work on any event loop. That's also why I'd like to rename the whole thing (see #50).

Alternatively, I can fork it as "anyio-mqtt". It's just that I really don't want to fragment this "async MQTT library" space any further. :))

**Reply by flyte** [reply]
Well, I read the articles you linked to and I'm sold on "structured concurrency". Not least because I've spent two days building and debugging a really small but disproportionately frustrating asyncio-based game server process manager with some MQTT cherries on top.

I really like the look of trio, not least because the documentation is _so_ much better than asyncio's. The trio implementation (trio-paho-mqtt) of what we're doing here is embarrassingly simple (albeit perhaps missing some of the features) and I "want me some of that". From what I've seen, AnyIO provides a similar API, so I'm enthusiastic about that too.

Since trio-paho-mqtt is so simple, I'd just start from scratch with anyio(-paho)-mqtt, which I suspect is what you did anyway, and then deprecate/replace this client once it reaches parity with it. That way you're only temporarily further fragmenting the MQTT client space.

**Reply by frederikaalund** [reply]
Glad to hear that. :) Let me see if I can find that anyio-based port I made a couple of months back.

**Comment by frederikaalund** [comment]
I made a work-in-progress anyio port in a new branch: [feat-anyio](https://github.com/sbtinstruments/asyncio-mqtt/tree/feat-anyio). This was the easiest for me; we can move to a new repo later.

See the [`anyio_example.py` file](https://github.com/sbtinstruments/asyncio-mqtt/blob/feat-anyio/anyio_example.py) for details.

**Comment by 0x0a11c0de** [comment]
My opinion is this library should only require asyncio. If it supports anyio, cool, but it should not require it.

**Reply by frederikaalund** [reply]
Hereby noted. :+1: Thanks for the feedback. I'll strive to make the anyio part optional so that you can use asyncio-mqtt as if it was just plain old asyncio. :)

**Comment by resingm** [comment]
If I am allowed to contribute to the discussion, I am opposed to the change. asyncio-mqtt makes it easy to write async MQTT applications with very few dependencies. In fact, it is just paho. I'd rather keep it that way. All other MQTT libs I found so far, are either not as mature and reliable or way larger than this lib.

**Reply by frederikaalund** [reply]
Of course, this is a call to everybody for input. The more feedback we get, the better. :+1: 

Indeed, anyio would be another dependency. That results in an overall "heavier" package. I share your concern here. Right now, asyncio-mqtt feels rather "light", which is a good thing. I'd hate to change that.

The maintainability of an anyio-based code base is just so much easier than a pure asyncio-based code base. I can't ignore that fact.  Maybe we should aim for an anyio-based fork instead. My concern is that this effectively deprecates asyncio-mqtt, as we focus the development effort on the fork.

In any case, thanks for the feedback. :+1:

**Comment by flyte** [comment]
I made [my own rough take](https://github.com/flyte/anyio-mqtt/blob/develop/anyio_mqtt.py) on the AnyIO Paho wrapper at the weekend. Just wanted to use it as a learning experience and I needed something for one of my projects.

Now, I've registered anyio-mqtt on PyPI just so we have the name, but I'm absolutely not planning on camping on it. My interest purely lies in ending up with a decent AnyIO MQTT client, so whatever ends up being the best one can live there. If anybody starts seriously using my version from PyPI even though it has no readme/description then that's sort of on them :sweat_smile:

**Reply by flyte** [reply]
During creation of this, it occurred to me that it's quite possibly impossible to wrap Paho in a way that doesn't use threads unless Paho can allow us to handle socket creation ourselves. I also encountered quite a few issues with deadlocks thanks to all the mutexes used by Paho. Of course I fantasised about writing my own client (possibly using the Paho C library to handle the protocol, or just implementing the MQTT v5 myself), but it's not really a project I can dedicate enough time to.

**Reply by frederikaalund** [reply]
Great to hear that. :+1: I can barely find the time to keep asyncio-mqtt up to date, so feel free to take ownership of anyio-mqtt. I can help co-maintain if you want. :)

> it occurred to me that it's quite possibly impossible to wrap Paho in a way that doesn't use threads unless Paho can allow us to handle socket creation ourselves

Yeah, that's a shortcoming in paho-mqtt at the moment. Alternatively, we can add a non-blocking way to create the socket in paho-mqtt. My only gripe about that approach is that the paho maintainers [do not seem so interested in my pull requests](https://github.com/eclipse/paho.mqtt.python/pull/530). Maybe others have more luck.

> Of course I fantasised about writing my own client (possibly using the Paho C library to handle the protocol, or just implementing the MQTT v5 myself), but it's not really a project I can dedicate enough time to.

Sounds like a fun project. :) I can relate to the time aspect of it, though. :sweat_smile:

**Comment by agronholm** [comment]
I, too, found this in search of an AnyIO based MQTT client. I took a look at several alternatives, and asyncio-mqtt had by far the best looking API. The ideal client library would be one with this API but with asynchronous internal machinery that uses AnyIO's APIs. I'm surprised nobody has create a sans-io implementation of MQTT (any version) for Python. That would be even better!

I would do this work myself, but I have my hands full at the moment busy with AnyIO itself, APScheduler and other projects that have not had the attention they deserve. Perhaps something like [smtpproto](https://github.com/agronholm/smtpproto) could serve as a basis? The MQTT protocol, even v5, isn't _that_ complicated.

If I do this, would you relinquish the `anyio-mqtt` name @flyte ?

**Reply by frederikaalund** [reply]
Thanks for the input. I'm glad to hear that you like the API side of things. 👍 If anyone wants to combine the asyncio-mqtt API with an anyio-backend, be my guest. :) Make sure to include the [breaking API change from #48](https://github.com/sbtinstruments/asyncio-mqtt/issues/48#issuecomment-788866770). See the [experimental anyio branch for an example of this](https://github.com/sbtinstruments/asyncio-mqtt/blob/feat-anyio/anyio_example.py).

I agree that a sans-io approach is the way to go. Takes a lot of development and maintenance to reach feature- and stability-parity with paho-mqtt, though. Personally, I don't have the resources to do this. :)

Are you aware of https://github.com/M-o-a-T/distmqtt?

Thanks again for your input. 👍

**Reply by agronholm** [reply]
> Are you aware of https://github.com/M-o-a-T/distmqtt?

I am, and it was the first MQTT client I looked at, but while the author of that library is an early AnyIO adopter, I take issue with the general quality of his code which I often find not meeting my own standards. Besides, distmqtt only supports up to MQTT 3.1.

**Reply by norwood867** [reply]
agronholm, i'm curious about what you needed up using for a concurrent mqtt client?

**Reply by agronholm** [reply]
What do you mean by "concurrent" mqtt client?

**Reply by flyte** [reply]
> If I do this, would you relinquish the `anyio-mqtt` name @flyte ?

Apologies for not seeing this sooner. 

I'm happy to relinquish the name for a good cause! I'm not actively developing it now anyway due to having other mouths to feed with code :)

**Reply by norwood867** [reply]
> What do you mean by "concurrent" mqtt client?

mqtt library that utilizes the async concurrency library like Trio or AnyIO. Asyncio-mqtt is solid from what I can tell, but my stuff is built on the trio, and my hacked-together mqtt access could be more robust.

**Reply by agronholm** [reply]
> > What do you mean by "concurrent" mqtt client?
> 
> mqtt library that utilizes the async concurrency library like Trio or AnyIO. Asyncio-mqtt is solid from what I can tell, but my stuff is built on the trio, and my hacked-together mqtt access could be more robust.

Concurrency is not the point, but asynchronous operation is. The rest of the app is async, and the async programming model has the benefit of thread safety (by virtue of using a single thread for everything) and the ability to cancel operations, which blocking sockets don't have. Additionally, any calls to blocking functions need to be wrapped in threads, using `asyncio.to_thread()` or similar. This can be problematic in high traffic applications that often require threads for other things and which can exhaust the number of threads. This unfortunately happens frequently in our app.

**Comment by frederikaalund** [comment]
I have a feeling that we'll see an [anyio-like interface within asyncio in the standard library](https://github.com/python/cpython/blob/main/Lib/asyncio/taskgroups.py) in a future Python release. It's not everything that anyio is (maybe we'll never get all of it in the standard library) but it's certainly a great move towards structured concurrency.

**Reply by agronholm** [reply]
I discussed this last week with Guido van Rossum, who is just about the only asyncio maintainer left, and the consensus was that asyncio won't get any esoteric new features anymore, and people should perhaps just migrate to something else, like Trio. Of course, given that Trio's compatibility with asyncio stuff is a bit iffy (trio-asyncio has some jagged edges), AnyIO offers more comfortable middle ground.
